File: tests/performance/drivers/jdbc/app/src/net/snowflake/perf/Results.java
package net.snowflake.perf;

import java.io.IOException;
import java.io.Writer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
// Jackson is bundled in the fat jar, relocated by shadowJar (see Config).
import net.snowflake.client.jdbc.internal.com.fasterxml.jackson.databind.ObjectMapper;
import net.snowflake.client.jdbc.internal.com.fasterxml.jackson.databind.node.ObjectNode;

/** Writes the per-iteration CSV, memory-timeline CSV, and run-metadata JSON under /results. */
final class Results {

  private static final Path RESULTS_DIR = Paths.get("/results");
  private static final ObjectMapper MAPPER = new ObjectMapper();

  private Results() {}

  static Path writeCsvResults(
      List<QueryExecution.IterationResult> results, String testName, String driverType)
      throws IOException {
    Path dir = testDir(testName, driverType);
    Files.createDirectories(dir);
    Path file = dir.resolve(testName + "_jdbc_" + driverType + "_" + epochSeconds() + ".csv");
    boolean hasConcurrent = !results.isEmpty() && results.get(0).workerCount > 0;
    try (Writer w = Files.newBufferedWriter(file, StandardCharsets.UTF_8)) {
      w.write("timestamp_ms,query_s,fetch_s,row_count,cpu_time_s,peak_rss_mb");
      if (hasConcurrent) {
        w.write(",worker_count,throughput_rows_s");
      }
      w.write("\n");
      for (QueryExecution.IterationResult r : results) {
        w.write(String.format("%d,%.6f,%.6f,%d,%.6f,%.1f",
            r.timestampMs, r.queryTimeS, r.fetchTimeS, r.rowCount, r.cpuTimeS, r.peakRssMb));
        if (hasConcurrent) {
          w.write(String.format(",%d,%.1f", r.workerCount, r.throughputRowsS));
        }
        w.write("\n");
      }
    }
    return file;
  }

  static Path writePutGetCsvResults(
      List<PutExecution.IterationResult> results, String testName, String driverType)
      throws IOException {
    Path dir = testDir(testName, driverType);
    Files.createDirectories(dir);
    Path file = dir.resolve(testName + "_jdbc_" + driverType + "_" + epochSeconds() + ".csv");
    try (Writer w = Files.newBufferedWriter(file, StandardCharsets.UTF_8)) {
      w.write("timestamp_ms,query_s,cpu_time_s,peak_rss_mb\n");
      for (PutExecution.IterationResult r : results) {
        w.write(String.format("%d,%.6f,%.6f,%.1f%n",
            r.timestampMs, r.queryTimeS, r.cpuTimeS, r.peakRssMb));
      }
    }
    return file;
  }

  static Path writeMemoryTimeline(
      List<ResourceMonitor.Sample> timeline, String testName, String driverType)
      throws IOException {
    if (timeline.isEmpty()) {
      return null;
    }
    Path dir = testDir(testName, driverType);
    Files.createDirectories(dir);
    Path file = dir.resolve(
        "memory_timeline_" + testName + "_jdbc_" + driverType + "_" + epochSeconds() + ".csv");
    try (Writer w = Files.newBufferedWriter(file, StandardCharsets.UTF_8)) {
      w.write("timestamp_ms,rss_bytes,vm_bytes\n");
      for (ResourceMonitor.Sample s : timeline) {
        w.write(String.format("%d,%d,%d%n", s.timestampMs, s.rssBytes, s.vmBytes));
      }
    }
    return file;
  }

  static void writeRunMetadata(String driverType, String driverVersion, String serverVersion)
      throws IOException {
    Path file = RESULTS_DIR.resolve("run_metadata_jdbc_" + driverType + ".json");
    if (Files.exists(file)) {
      return;
    }
    Files.createDirectories(RESULTS_DIR);
    ObjectNode metadata = MAPPER.createObjectNode();
    metadata.put("driver", "jdbc");
    metadata.put("driver_type", driverType);
    metadata.put("driver_version", driverVersion);
    metadata.put("runtime_language_version", System.getProperty("java.specification.version"));
    metadata.put("server_version", serverVersion);
    metadata.put("architecture", architecture());
    metadata.put("os", envOrDefault("OS_INFO", "Linux"));
    metadata.put("run_timestamp", epochSeconds());
    if ("universal".equals(driverType)) {
      metadata.put("build_rust_version", envOrDefault("BUILD_RUST_VERSION", "NA"));
    }
    Files.write(
        file,
        MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(metadata)
            .getBytes(StandardCharsets.UTF_8));
  }

  private static long epochSeconds() {
    return System.currentTimeMillis() / 1000;
  }

  private static Path testDir(String testName, String driverType) {
    String subdir = testName.endsWith("_record") ? "_record" : testName;
    return RESULTS_DIR.resolve(driverType).resolve(subdir);
  }

  private static String architecture() {
    String arch = System.getProperty("os.arch", "").toLowerCase();
    if (arch.equals("amd64") || arch.equals("x64") || arch.equals("x86_64")) {
      return "x86_64";
    }
    if (arch.equals("aarch64") || arch.equals("arm64") || arch.equals("armv8")) {
      return "arm64";
    }
    return arch;
  }

  private static String envOrDefault(String key, String def) {
    String v = System.getenv(key);
    return (v == null || v.isEmpty()) ? def : v;
  }
}

File: tests/performance/drivers/_shared/results.py
"""Results output for single-implementation perf drivers (sqlapi, adbc)."""

import csv
import json
import os
import sys
import time
from pathlib import Path


def write_csv_results(results, test_name, driver):
    timestamp = int(time.time())
    subdir = "_record" if test_name.endswith("_record") else test_name
    results_dir = Path("/results") / "universal" / subdir
    results_dir.mkdir(parents=True, exist_ok=True)
    filename = results_dir / f"{test_name}_{driver}_{timestamp}.csv"

    with open(filename, "w", newline="") as f:
        writer = csv.DictWriter(
            f,
            fieldnames=[
                "timestamp_ms",
                "query_s",
                "fetch_s",
                "row_count",
                "cpu_time_s",
                "peak_rss_mb",
            ],
        )
        writer.writeheader()
        for result in results:
            writer.writerow(
                {
                    "timestamp_ms": result["timestamp"],
                    "query_s": f"{result['query_time_s']:.6f}",
                    "fetch_s": f"{result['fetch_time_s']:.6f}",
                    "row_count": result.get("row_count", 0),
                    "cpu_time_s": f"{result['cpu_time_s']:.6f}",
                    "peak_rss_mb": f"{result['peak_rss_mb']:.1f}",
                }
            )
    return filename


def write_memory_timeline(memory_timeline, test_name, driver):
    if not memory_timeline:
        return None
    timestamp = int(time.time())
    subdir = "_record" if test_name.endswith("_record") else test_name
    results_dir = Path("/results") / "universal" / subdir
    results_dir.mkdir(parents=True, exist_ok=True)
    filename = results_dir / f"memory_timeline_{test_name}_{driver}_{timestamp}.csv"
    with open(filename, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["timestamp_ms", "rss_bytes", "vm_bytes"])
        writer.writeheader()
        for sample in memory_timeline:
            writer.writerow(
                {
                    "timestamp_ms": sample.timestamp_ms,
                    "rss_bytes": sample.rss_bytes,
                    "vm_bytes": sample.vm_bytes,
                }
            )
    return filename


def write_run_metadata(server_version, driver, driver_version):
    metadata_filename = Path("/results") / f"run_metadata_{driver}.json"
    if metadata_filename.exists():
        return
    metadata = {
        "driver": driver,
        "driver_type": "universal",
        "driver_version": driver_version,
        "runtime_language_version": f"{sys.version_info.major}.{sys.version_info.minor}",
        "server_version": server_version,
        "architecture": _get_architecture(),
        "os": os.environ.get("OS_INFO", "Linux"),
        "run_timestamp": int(time.time()),
        "build_rust_version": "NA",
    }
    metadata_filename.write_text(json.dumps(metadata, indent=2))


def _get_architecture():
    import platform

    machine = platform.machine().lower()
    if machine in ("amd64", "x64", "x86_64"):
        return "x86_64"
    if machine in ("aarch64", "armv8"):
        return "arm64"
    return machine
